tmpl/blog/mirage-entropy.md

```diff
@@ -0,0 +1,341 @@
+after observing only 624 outputs in a row.
+
+We can look for something approaching a "true" random sequence in the nature.
+The current agreement is that the nature of quantum processes is random in this
+sense, and random sequences based on this idea are readily available for
+[download][dgrb-boskovic]. Or we can use the microphone and keep recording; the
+lowest-order bits of the signal are pretty random. But we cannot write a program
+to generate an actually random sequence.
+
+Still, we need to compromise. Somewhere in between the basic jigglyness and a
+sequence that cannot be described by anything other than itself is the area of
+Cryptographically Secure Pseudo Random Number Generators (CSPRNGs). These have a
+property that it is computationally prohibitively expensive to distinguish their
+output from a "true" random sequence, with a high probability. This also means
+that it is computationally prohibitively expensive to reconstruct their internal
+state, just by looking at their outputs. In a sense, someone trying to predict
+the outputs can not take a shortcut and is reduced to starting the generator
+with all the possible states and checking if the output of their generator
+matches the observed sequence, a laborious task. This is how we can quantify a
+CSPRNG unpredictability: it takes trying about half of all the possible states to
+guess the state.
+
+MirageOS's security stack contains a CSPRNG, a design called [Fortuna][fortuna].
+What it really does, is encrypt the simple sequence `0, 1, 2, 3, ...` with AES
+(AES-CTR) using a secret key. This makes it as resistant to prediction as AES is
+to [known-plaintext attacks][wiki-known-plaintext]. After each output, it
+generates a bit more, hashes that, and uses the result as the next key. This is
+not to improve the statistical randomness, as this is already guaranteed by AES.
+Rather, it's a form of [forward secrecy][wiki-forward-secrecy]: an attacker who
+learns the secret key at some point would need to perform the [preimage
+attack][wiki-preimage-attack] on the hash function to figure out the earlier key
+and reconstruct the earlier outputs.
+
+[wiki-random-sequence]: https://en.wikipedia.org/wiki/Random_sequence
+[wiki-statistical-randomness]: https://en.wikipedia.org/wiki/Statistical_randomness
+[diehard-web]: http://www.stat.fsu.edu/pub/diehard/
+[marsaglia-pi-randomness]: http://interstat.statjournals.net/YEAR/2005/articles/0510005.pdf
+[wiki-algorithmic-randomness]: https://en.wikipedia.org/wiki/Algorithmically_random_sequence
+[wolfram-rule-30]: http://mathworld.wolfram.com/Rule30.html
+[mersenne-twister]: https://en.wikipedia.org/wiki/Mersenne_Twister
+[dgrb-boskovic]: http://qrbg.irb.hr/
+[fortuna]: https://www.schneier.com/fortuna.html
+[wiki-known-plaintext]: https://en.wikipedia.org/wiki/Known-plaintext_attack
+[wiki-forward-secrecy]: https://en.wikipedia.org/wiki/Forward_secrecy
+[wiki-preimage-attack]: https://en.wikipedia.org/wiki/Preimage_attack
+
+## Entropy
```

Comment:
I think this part of the article would benefit of making a clear distinction between `RNG` and `PRNG`. This is true of the whole article, most of the `RNG` acronyms should be replaced `PRNG`. This `pseudo` is important.